### DISCUSSION TITLE
Using _total_rotation to rotational transformation of submartices of beam element stiffness matrix

### DISCUSSION THREAD

**Discussion by andrey1887** [idaholab/moose#26400] [Q&A Modules: Solid mechanics]
Hello. In `ComputeIncrementalBeamStrain::computeStiffnessMatrix()` the next code is present:
```
  // relation between translational displacements at node 0 and translational forces at node 0
  RankTwoTensor K11_local;
  K11_local.zero();
  K11_local(0, 0) = youngs_modulus * A_avg / _original_length[0];
  K11_local(1, 1) = shear_modulus * A_avg / _original_length[0];
  K11_local(2, 2) = shear_modulus * A_avg / _original_length[0];
  _K11[0] = _total_rotation[0].transpose() * K11_local * _total_rotation[0];

  // relation between displacements at node 0 and rotational moments at node 0
  RankTwoTensor K21_local;
  K21_local.zero();
  K21_local(2, 1) = shear_modulus * A_avg * 0.5;
  K21_local(1, 2) = -shear_modulus * A_avg * 0.5;
  _K21[0] = _total_rotation[0].transpose() * K21_local * _total_rotation[0];

  // relation between rotations at node 0 and rotational moments at node 0
  RankTwoTensor K22_local;
  K22_local.zero();
  K22_local(0, 0) = shear_modulus * Ix_avg / _original_length[0];
  K22_local(1, 1) = youngs_modulus * Iz_avg / _original_length[0] +
                    shear_modulus * A_avg * _original_length[0] / 4.0;
  K22_local(2, 2) = youngs_modulus * Iy_avg / _original_length[0] +
                    shear_modulus * A_avg * _original_length[0] / 4.0;
  _K22[0] = _total_rotation[0].transpose() * K22_local * _total_rotation[0];

  // relation between rotations at node 0 and rotational moments at node 1
  RankTwoTensor K22_local_cross = -K22_local;
  K22_local_cross(1, 1) += 2.0 * shear_modulus * A_avg * _original_length[0] / 4.0;
  K22_local_cross(2, 2) += 2.0 * shear_modulus * A_avg * _original_length[0] / 4.0;
  _K22_cross[0] = _total_rotation[0].transpose() * K22_local_cross * _total_rotation[0];

  // relation between displacements at node 0 and rotational moments at node 1
  _K21_cross[0] = -_K21[0];
```
So for all submatrices `_K11`, `_K21`, `_K22`, `_K21_cross`, `_K22_cross` the `_total_rotation` applied, this mean, that each 3x3 part of stiffness matrix will be transformed the same way.

In some literature sources i found the next form of transformation matix:
![MicrosoftTeams-image (8)](https://github.com/idaholab/moose/assets/114737834/78cf6589-be51-49d5-9dd1-4453406a0384)

Explain me please why in current beam element implementation the different approach is used.

**Comment by GiudGiud** [comment]
@cbolisetti

**Reply by cbolisetti** [reply]
The literature snippet you showed is for state variables - displacements, rotations, etc. The way to rotate stiffness matrices from global to local is different. If you derive that for the stiffness matrices, you get what you see in the code.

**Reply by andrey1887** [reply]
Thank you @cbolisetti and @GiudGiud.
Another example:
![MicrosoftTeams-image (10)](https://github.com/idaholab/moose/assets/114737834/3556df30-747d-405c-9076-018ba0698666)

**Reply by cbolisetti** [reply]
Yes, this is consistent with the code in MOOSE.

**Reply by andrey1887** [reply]
Thank you, @cbolisetti, but i can't understand how it can be. T matrix consist of 4 submartices placed on it's diagonal, all another elements equals to 0. It is the same thing as in my first message.

**Reply by andrey1887** [reply]
@cbolisetti , i have some simply example of c++ code fore 3d truss element.
[finite-element-method-by-example-in-qt-slash-c-plus-plus (3).pdf](https://github.com/idaholab/moose/files/13755166/finite-element-method-by-example-in-qt-slash-c-plus-plus.3.pdf)
And there i see, that `T` for the force vector transformation and `T` for stiffness matrix transformation are the same - in (21) and (29).
![MicrosoftTeams-image (11)](https://github.com/idaholab/moose/assets/114737834/e8055b55-2da4-4da2-85cf-e4a6d5d9351c)
May be you meant the final variant of single transformation matrix which multiplied on `k` in (29):
![изображение](https://github.com/idaholab/moose/assets/114737834/de4dff3b-3802-44ef-ba16-9f1052e46566)
Then i agree that they are same, but in moose beam element used both transposed and initial `_total_rotation`:
` _K22_cross[0] = _total_rotation[0].transpose() * K22_local_cross * _total_rotation[0];`

**Reply by cbolisetti** [reply]
Are you talking about the difference between having 4 us matrices in the diagonal vs multiplying them individually? I haven't done the math but I trust the developers that they are equivalent. I'd be surprised if they aren't. May be you can try doing the math separately and let us know if the implementation is actually wrong.

**Reply by andrey1887** [reply]
Hello, @cbolisetti , sorry for late response. I did the next:
1. First i consider, the local stiffness matrix `K` from 3.66 have elements equals to one and multiple `T` matrix to `K`:
![1st html](https://github.com/idaholab/moose/assets/114737834/b58b9398-c2a7-4c21-ac32-71c54ceda857)

2. After that i took obtained result and multiply it to T (transposed):
![2nd html](https://github.com/idaholab/moose/assets/114737834/7faad879-1cbc-45f2-8b47-c3a06be3e95d)

3. Finally this results in:
![math html](https://github.com/idaholab/moose/assets/114737834/752aa6f2-cb74-45d9-91cf-d5a2c5e39039)

Where highlighted snippet (3х3) is repeated 16 times.
So you were right, that multiplying whole matrices and multiplying it's submatrices with subsequent substitution into overall are equivalent.
Thank you.
